targets/TARGET_STM/hal_tick_overrides.c

```diff
@@ -47,26 +46,28 @@ HAL_StatusTypeDef HAL_InitTick(uint32_t TickPriority)
 
 uint32_t HAL_GetTick()
 {
-#if TIM_MST_BIT_WIDTH == 16
-    uint32_t new_time;
-    if (mbed_sdk_inited) {
-        // Apply the latest time recorded just before the sdk is inited
-        new_time = ticker_read_us(get_us_ticker_data()) + prev_time;
-        prev_time = 0; // Use this time only once
-        return (new_time / 1000);
+    uint32_t new_time = us_ticker_read();
+    uint32_t elapsed_ticks = 0;
+    uint32_t elapsed_time = (((new_time - prev_time) & US_TICKER_MASK) + prev_tick_remainder);
+    prev_time = new_time;
+    if (elapsed_time < 1000) { // elapsed time less than 1ms
+        elapsed_ticks = 0;
+        prev_tick_remainder = elapsed_time;
+    } else if (elapsed_time < 2000) { // elapsed time less than 2ms
+        elapsed_ticks = 1;
+        prev_tick_remainder = elapsed_time - 1000;
+    } else if (elapsed_time < 65536) { // For 16-bit ticker max value of 2^16 = 65536
```

Comment:
The maximum `elapsed_time` value here is actually 65535+999, but I don't think that's a relevant factor for the threshold.

We just want to not spend a huge time going round the while loop, when the C division would be faster.

That crossover point would be pretty low for most C libraries, but maybe not microlib.

I had a vague idea the compiler could optimise out the test for a 16-bit timer if it was set high though - a space saving rather than a speed saving.

Maybe that would work, if we help it:

    if `(TIM_MST_BIT_WIDTH <= 16 || elapsed_time < 65536)`

That will basically force the compiler to leave out the `/%` case when 16-bit.